bbknn/matrix.py

```diff
@@ -190,40 +217,19 @@ def get_graph(pca, batch_list, params):
         (pca.shape[0], params["neighbors_within_batch"] * len(batches))
     )
     knn_indices = np.copy(knn_distances).astype(int)
-    # find the knns using faiss/cKDTree/KDTree/annoy
-    # need to compare each batch against each batch (including itself)
-    for to_ind in range(len(batches)):
-        # this is the batch that will be used as the neighbour pool
-        # create a boolean mask identifying the cells within this batch
-        # and then get the corresponding row numbers for later use
-        batch_to = batches[to_ind]
-        mask_to = batch_list == batch_to
-        ind_to = np.arange(len(batch_list))[mask_to]
-        # create the faiss/cKDTree/KDTree/annoy, depending on approx/metric
-        ckd = create_tree(data=pca[mask_to, : params["n_pcs"]], params=params)
-        for from_ind in range(len(batches)):
-            # this is the batch that will have its neighbours identified
-            # repeat the mask/row number getting
-            batch_from = batches[from_ind]
-            mask_from = batch_list == batch_from
-            ind_from = np.arange(len(batch_list))[mask_from]
-            # fish the neighbours out, getting a (distances, indices) tuple back
-            ckdout = query_tree(
-                data=pca[mask_from, : params["n_pcs"]], ckd=ckd, params=params
-            )
-            # the identified indices are relative to the subsetted PCA matrix
-            # so we need to convert it back to the original row numbers
-            for i in range(ckdout[1].shape[0]):
-                for j in range(ckdout[1].shape[1]):
-                    ckdout[1][i, j] = ind_to[ckdout[1][i, j]]
-            # save the results within the appropriate rows and columns of the structures
-            col_range = np.arange(
-                to_ind * params["neighbors_within_batch"],
-                (to_ind + 1) * params["neighbors_within_batch"],
-            )
-            knn_indices[ind_from[:, None], col_range[None, :]] = ckdout[1]
-            knn_distances[ind_from[:, None], col_range[None, :]] = ckdout[0]
-    return knn_distances, knn_indices
+    batch_combos = list(itertools.product(batches, batches))
+
+    args = [(b[0], b[1], pca, batch_list, params) for b in batch_combos]
+
+    with multiprocessing.Pool() as pool:
+        results = pool.starmap(_get_indices, args)
```

Comment:
how about:

```python
 for result in tqdm(
                pool.imap_unordered(_get_indices, args),
                total=len(args),
                disable=(not progress),
            ):
                results.append(result)
```

where `progress` is a boolean whether to show a progress bar? And then you'll need to change `_get_indices()` to accept a single argument and expand the input tuple.